language/symbol-pool/src/lib.rs

```diff
@@ -0,0 +1,41 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+// NOTE: If you're looking for a forbid(unsafe_code) attribute here, you won't
+// find one: symbol-pool (and its inspiration, servo/string-cache) uses unsafe
+// code in order to store and dereference the memory addresses of strings.
+
+use once_cell::sync::Lazy;
+use serde::{Deserialize, Serialize};
+use std::{collections::HashSet, num::NonZeroU64, ptr::NonNull, sync::Mutex};
+
+static SYMBOL_POOL: Lazy<Mutex<HashSet<String>>> = Lazy::new(|| Mutex::new(HashSet::new()));
```

Comment:
`Box<str>` at the very least here, but any reason to not just go ham with a larger fork of `string_cache`?

I'd probably just start by more or less copy-pasting the `string_cache` code and then start ripping out all of the ref-counting related items

Reply:
Then again, maybe we can:
- Start here
- Move everything in move-lang to using this (basically making `Name` a `Spanned<symbol>` instead of `Spanned<string>` 
- Then actually verify perf changes?

Reply:
I think starting small would be my personal preference -- definitely defer to your judgement here if you're saying it's guaranteed we'll need more of the optimizations from string_cache, but my take is that we can always plug those bits in later. And, like you mentioned, verifying the perf improvements might be easier that way (I guess we could add it all now, then comment out the inlining parts later, but that seems more complex).

Reply:
I think the optimizations are rather easy to add, and it will be fun to actually see what the affect is. So I'm good with this!